an-introduction-to-unit-testing.md

```diff
@@ -0,0 +1,852 @@
+
+### Test-driven development (TDD)
+
+TDD is a practice that involves writing unit tests before the production code is written, writing a failing test to prove that code or functionalities are missing, and then making the test pass by writing production code that meets the expectations.
+
+When refactoring your code (refactoring means changing a piece of code without changing its functionality), seeing a test fail, and then seeing it pass without changing the test, you’re basically testing the test itself too.
+
+### Types of unit testing
+
+**Value-based testing** checks the value returned from a function.
+
+**State-based testing** is about checking for noticeable behavior changes.
+
+**Interaction testing** is testing how an object sends messages (i.e. calls methods) to other objects. You use interaction testing when calling another object is the end result of a specific unit of work.
+
+## Unit testing in practice
+
+This section introduces the xUnit testing frameworks.
+
+### Unit testing frameworks
+
+Unit tests are written as code, using libraries from the unit testing framework. The tests are run from a separate unit testing tool or inside the IDE, and the results are reviewed (either as output text, the IDE, or the unit testing framework application UI) by the developer or an automated build process.
+
+### The xUnit frameworks
+
+xUnit is a collective name for several unit testing frameworks that derive their structure and functionality from Smalltalk’s SUnit, designed by Kent Beck in 1998.
+
+### Naming convention
+
+The name of a unit test usually includes the name of the class, the name of the unit of work, and the expected behavior.
+
+Example:  `OpticalSensor_IsRedBloodCell_ReturnsFalse()`
+
+### Structure of a unit test
+
+The typical sequence of actions performed by a unit test is the following: creating an object, setting it up as necessary, acting on the object, and asserting that something is as expected.
+
+### Attributes
+
+The unit testing framework provides several attributes used to decorate the code of the unit tests, such as  `TestFixture`,  `Test`,  `TestCase`,  `Setup`, and  `TearDown`.
+
+### Assertions
+
+The unit testing also provides classes and methods to assert what is expected, such as  `Assert.IsFalse()`,  `Assert.IsTrue()`,  `Assert.AreEqual()`,  `Assert.Throws()`, and  `Assert.DoesNotThrow()`.
+
+## Fakes
+
+This section presents faking techniques commonly used in unit testing.
+
+### External dependencies
+
+An external dependency is an object in your system that your code under test interacts with and over which you have no control.
+
+### Fakes: stubs and mocks
+
+A fake is a generic term that can describe either a stub or a mock. A stub is a controllable replacement for an existing dependency: by using a stub, you can test your code without dealing with the dependency directly. Mocks are just like stubs, but you can assert against the mock object, whereas you do not assert against a stub.
```

Comment:
I'm not sure, but I think mocking framework such as Nsubstitute or Moq are also mentionable and maybe as an extension to that some AutoMock addon to those frameworks. Because it reduces a lot of maintenance effort in your tests. But that's up to you.